### ISSUE TITLE
Making Minetest easy to install on all operating systems

### ISSUE THREAD

**Issue by Calinou** [minetest/minetest#3596]
Discussions on IRC and forums have shown that Minetest is not that easy to install and integrate with the system, so I'm making this issue to keep track of problems related to ease of installation.

What needs to be done:
- [ ] In portable Windows builds, put the `minetest.exe` binary at the root rather than in `bin/`
- [ ] Provide official Windows installers ([NSIS](http://nsis.sourceforge.net/Main_Page)? [InnoSetup](http://www.jrsoftware.org/isinfo.php)? [Squirrel](https://github.com/Squirrel/Squirrel.Windows)?)
  - [ ] 64-bit
  - [ ] 32-bit
- [ ] Provide official, distribution-independent GNU/Linux binaries (Static linking? `.so` bundling? [AppImage](3568)?)
  - [ ] 64-bit
  - [ ] 32-bit
- [ ] Provide official OS X `.dmg` or `.pkg`
  - [ ] 64-bit

**Comment by kaeza** [comment]
For the portable Win32, I'd add a simple `Play.bat` (or `Play.lnk` [Windows shortcut file]) running `bin\minetest.exe`. I like that the binaries (executables *and* DLLs) are kept in their own directory, not cluttering the root.

**Comment by rubenwardy** [comment]
I have made 64bit Windows build using Inno Setup. I'll make a 32bit one now.

**Comment by Calinou** [comment]
> For the portable Win32, I'd add a simple Play.bat (or Play.lnk [Windows shortcut file]) running bin\minetest.exe. I like that the binaries (executables and DLLs) are kept in their own directory, not cluttering the root.

Why not just `minetest.bat` then? At least, it doesn't lose all its meaning when you take it out of context.

**Comment by rubenwardy** [comment]
Added a windows 32 bit installer to the releases. Updating website now...

**Comment by kaeza** [comment]
> Why not just minetest.bat then? At least, it doesn't lose all its meaning when you take it out of context.

That's what I said, isn't it?

**Comment by rubenwardy** [comment]
It you move it out of it's folder for some reason, maybe accidentally.

**Comment by red-001** [comment]
It can't be *too* hard to change some settings to make minetest run from root of the minetest folder? No need for a .bat file. It also means minetest doesn't loss it's icon.

**Comment by rubenwardy** [comment]
Nothing to do with setting: on linux it works when you run it from the root folder using ./bin/minetest.

**Comment by rubenwardy** [comment]
It works already when you run it from the root folder using ./bin/minetest.

**Comment by kaeza** [comment]
>It you move it out of it's folder for some reason, maybe accidentally.

Then how would it find `minetest.exe`?

**Comment by rubenwardy** [comment]
It probably wouldn't. If it was accidentally moved then it wouldn't. I don't really know what Calinou meant.

**Comment by rubenwardy** [comment]
Personally I'd prefer a shortcut over a bat.

**Comment by red-001** [comment]
How do you make a shortcut to a portable file? I don't think you can. I thought @Calinou meant running minetest.exe from root as in what freeminer does. Isn't this just a cmake setting?

**Comment by kaeza** [comment]
>Personally I'd prefer a shortcut over a bat.

It would be ideal, partly because it allows you to set an icon for it (though I'm not really sure if icons work with relative paths), and partly because it hides the file extension regardless of the "Show extension for know file types" setting.

**Comment by rubenwardy** [comment]
I think shortcuts can be relative? I meant a shortcut in minetest root to the file in bin.

**Comment by red-001** [comment]
why not move the file?

**Comment by kaeza** [comment]
>why not move the file?

If you move `minetest.exe`, you must move all the DLLs, cluttering the root.

**Comment by red-001** [comment]
you could make it search for dlls in /bin, right?

**Comment by kaeza** [comment]
>you could make it search for dlls in /bin, right?

Yes, but you would need to alter `PATH` or similar, **before** running `minetest.exe`. So let's put `minetest.exe` in the root, along with, dunno, a batch file or a shortcut that modifies `PATH`, and then runs `minetest.exe` in the root. Sooo... why did we move `minetest.exe` to the root in the first place?

A batch file or shortcut does the job nicely: does not require patching the engine (if needed) to look in e.g. `$exepath/games` (currently it looks in `$exepath/../games` IIRC, which works from `bin/`), and keeps everything mostly tidy, only requiring a single extra file in the root. Am I missing something here?

**Comment by red-001** [comment]
Relative path shortcuts on windows are at best a hack, they can't have the minetest icon because they work by making command prompt start the program. `.bat` files can't have an icon either. If you don't want to make a few small changes to minetest engine then the best you can do will being more user friendly is to make another executable that will work as a launcher for minetest.

**Comment by est31** [comment]
`Relative path shortcuts on windows are at best a hack, they can't have the minetest icon because they work by making command prompt start the program`

Windows does support both softlinks and hardlinks to my knowledge, at least since vista I think. They just don't have an UI in the explorer.

**Comment by Dragonop** [comment]
Just adding this: you can't attach a .bat shortcut in the task bar in Windows

**Comment by kaeza** [comment]
>[...]they can't have the minetest icon because they work by making command prompt start the program.

What do you mean by "command prompt", and what does it have to do with shortcuts? Did you mean `.bat` files?

>[...]the best you can do will being more user friendly is to make another executable that will work as a launcher for minetest.

That could be a viable solution, even if the "launcher" just `execv`s `bin/minetest`.

@est31 Yep, Windows does support links as of Vista, but IIRC, from the point of the program loader, you still have the executable where the path is.

Maybe I'm misunderstanding @Calinou's point about this. I guess his point is that `minetest.exe` is "hidden" in `bin`. I'd add that it is also mixed among several DLLs that may confuse/overwhelm non-tech-savvy users.

**Comment by est31** [comment]
Why not just use LNK files then?

**Comment by kaeza** [comment]
@est31 I was wrong. Creating `.lnk` files with relative targets is not possible (at least from the GUI; it may or may not be possible programatically).

**Comment by est31** [comment]
@kaeza : http://stackoverflow.com/questions/3418541/how-do-i-create-a-shortcut-lnk-with-a-relative-target

**Comment by kaeza** [comment]
@est31 Good to know it can be done, but this line discards that option:

>The icon is problematic, does not seem like it works with relative paths [...]

I agree with @red-001. A simple frontend/launcher `.exe` would be better.

**Comment by rubenwardy** [comment]
The installer fixes most of these problems.

**Comment by kaeza** [comment]
>The installer fixes most of these problems.

Except the "no need for an installer" problem, which is, the *Whole. Fucking. Point.* of a portable package.

Why do I even bother anymore?

**Comment by Calinou** [comment]
> I agree with @red-001. A simple frontend/launcher .exe would be better.

A launcher typically means you must go through additional clicks. It does not improve the system integration of Minetest at all, eg. by providing system-wide shortcuts available in the Start Menu. (Currently, you can create one by hand and place it in `%ProgramData%\Microsoft\Windows\Start Menu\Programs\Minetest`).

**Comment by gjgfuj** [comment]
@Calinou a launcher isn't neccessarily additional clicks, if it's just a simple .exe that launches it without any kind of GUI or anything. all it'd need to be is a quick wrapper program.

**Comment by rubenwardy** [comment]
There's no need for that tone, @kaeza. Very rude.
If the point is to be noob friendly, then you should use an installers. Note that I said most of these problems.

> A launcher typically means you must go through additional clicks. It does not improve the system integration of Minetest at all, eg. by providing system-wide shortcuts available in the Start Menu. (Currently, you can create one by hand and place it in %ProgramData%\Microsoft\Windows\Start Menu\Programs\Minetest).

Alll the installer does is unzip some files then make a shortcut. If you want system integration, that's how you do it.

**Comment by lugarius** [comment]
http://www.portablelinuxgames.org/game/minetest

i can contact the maker... he can teach you, he helped me too 

tux AT portablelinuxgames.org

**Comment by lugarius** [comment]
Other appImage help. 

https://github.com/RazZziel/PortableLinuxGames/wiki

**Comment by lugarius** [comment]
Another option is just include ALL libraries into an folder called lib... and let the executable take them and not from the system...

See www.xonotic.org they have done it already, appimage seems to need more time

**Comment by C1ffisme** [comment]
@lugarius Please edit your previous comments rather than posting several times.

**Comment by est31** [comment]
Also @lugarius Appimage itself is discussed in #3568 in detail, so better to discuss about it there.

**Comment by tasgoon** [comment]
I also noticed this issue, and to combat this, I have started work on a launcher for the game, similar to MultiMC and the default Minecraft Launcher. You can find it [here](https://github.com/tasgoon/MinetestLauncher). Right now it only works for Linux, and compiles each new instance from source. The plan is to download the releases straight from GitHub Releases for Windows, but I want to make sure Linux works relatively well before moving on to Windows.

Would this be a suitable solution for the issue at hand?

**Comment by est31** [comment]
That's cool!

There might be a chicken-egg-problem for the use case of getting minetest releases work on all different distros, but it will be very useful for non technical users who want to try latest minetest from git.

**Comment by tasgoon** [comment]
It actually doesn't compile from master (though that's easy to add). It uses the GitHub tags to allow the user to download and compile any tagged version of Minetest. Unfortunately, it's limited to compiling, as Minetest doesn't actually provide releases for Linux (though that's kind of what this entire issue is about).

**Comment by benrob0329** [comment]
FYI I can compile Minetest in < 5 minutes, so I'd be happy to compile and cross compile stuff.

**Comment by alugarius** [comment]
Any succsess in Appimage??

**Comment by Calinou** [comment]
> Any succsess in Appimage??

I have tried making a quick-and-dirty AppImage of the latest Minetest Git on a Ubuntu 12.04 64-bit chroot, but it does not work outside of the chroot (on Arch Linux, for example), I believe it needs some tweaking. It also doesn't provide LevelDB support due to a bug in Ubuntu 12.04.

**Comment by sofar** [comment]
With flatpak being usable, it may be more interesting to go that route for Linux.

**Comment by alugarius** [comment]
Try Ubuntu 14.xx LTS..

**Comment by est31** [comment]
I've created a flatpak build script for the minetest client: https://github.com/est31/minetest-flatpak

**Comment by est31** [comment]
Maybe we can even make an official repo for flatpak for release + nightly, but someone would need to supply a buildserver for this (hosting can happen on github pages).

**Comment by alugarius** [comment]
@Calinou 
https://forum.minetest.net/viewtopic.php?f=42&t=15802

I am for AppImage, add them to the download page?

**Comment by Luraktinus** [comment]
any news?

**Comment by paramat** [comment]
Do the checkboxes need updating?

**Comment by lucasburlingham** [comment]
Here is an installer package for Minetest 0.4.16. I can create more for later versions if necessary.  https://github.com/lucasburlingham/Minetest-0.4.17.1-Client/raw/macosx/minetest-0.16.0-client-macosx.pkg

**Comment by paramat** [comment]
Is there core dev support for the remaining checkboxes? Or should this be closed?

**Comment by Calinou** [comment]
Now that we have Windows installers and a [Minetest Flatpak](https://flathub.org/apps/details/net.minetest.Minetest), I think only macOS support remains an issue. However, we don't have any dedicated maintainer for it :slightly_frowning_face:

So, feel free to close this issue.

**Comment by rubenwardy** [comment]
We don't have Windows installers - having the capability doesn't count if they're not created

**Comment by Calinou** [comment]
@rubenwardy I just checked the website and can't see them on the Download page indeed. I remember them being there at some point.

**Comment by LoneWolfHT** [comment]
I have a Windows boot. I'm willing to create Windows installers for each release (Or figuring out a Github-run script).
My main OS is Linux, so I can look into creating AppImages too if @benrob0329 doesn't want to do that

**Comment by LoneWolfHT** [comment]
Hmmm, it looks like we already have the ability to make installers? Why aren't we using it?

**Comment by Tcll** [comment]
still no portable linux download (.tar.gz, .appimage)??

**Comment by themagicalcats** [comment]
> still no portable linux download (.tar.gz, .appimage)??

Please, do this, I beg.

**Comment by Calinou** [comment]
@Tcll @themagicalcats Please don't bump issues without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by Tcll** [comment]
> I believe the Flatpak is a good enough solution to installing Minetest on any Linux distribution nowadays.

flatpak requires internet to use apps and is not portable
the PC I can play minetest on is offline

this is why I suggested an appimage or pre-built tar.gz I can transfer to that machine
all I'll have to do is extract and run (appimage doesn't need to be extracted, but I do so anyways for dev reasons)

it's up to the devs to build binaries for users, not users to build binaries for themselves

I should also note I'm against system installations for portable apps, which is why I don't just download the xbps package
I prefer my portable apps on my secondary HDD for use with multiple linux distros

**Comment by jurf** [comment]
Only stumbled on this by accident, so I’ll just pipe in:

> flatpak requires internet to use apps

This isn’t true. The whole infrastructure for offline distribution is in place and has been actively used for years. See [the docs](https://docs.flatpak.org/en/latest/single-file-bundles.html) for simple instructions.

If you need portability (aka you want it to be able to be run from a USB), then also provide an AppImage. Although I’d argue that’s a separate use-case.

**Comment by Tcll** [comment]
> The whole infrastructure for offline distribution is in place and has been actively used for years.

alright thanks, I had some bad experiences in the past, particularly with SparkleShare, but there were a few others as well...

> I’d argue that’s a separate use-case.

portability is the primary reason appimage exists, it's anything BUT a separate use case ;)

but I'm suggesting tar.gz as well like Blender does in case appimage is too much of a hassle to build properly
just build the app and throw it in a tar.gz for the same reason you'd build an appimage

**Comment by benrob0329** [comment]
I had some success building a Minetest appimage some time ago, but was waiting on a few tooling changes to make it easier on older distros (namely pulling in the appdata/appinfo/what'cha'ma'call'it parser from an appimage so-as no to rely on the outdated one in the repos).

Those changes may be done now, though.

**Comment by Tcll** [comment]
just wanted to update, with help from a friend, I've found these which seem to work quite well :)
https://github.com/An0n3m0us/Minetest-AppImages/releases
for an added advantage, I don't even have to extract to add a missing lib, it just works as is :D 

it's a shame these aren't official though u.u

**Comment by jurf** [comment]
@Tcll I meant portability is a separate use-case to offline installation. I still think Flatpak is better for an offline computer because then you’d probably have more apps installed and it makes it easier to keep them updated and it uses disk space better. But an AppImage is vastly superior to carrying it around on say a USB stick.

**Comment by Tcll** [comment]
> I meant portability is a separate use-case to offline installation.

oh I see, portability vs offline installation, I suppose I can agree with that >.>

though I still disagree updates are easier to manage through an automated interface
if anything, it's more a hassle when you want to manage multiple versions of the same app
eg: I still have a need for collada importers on Blender 2.49 because the 2.79b importers are broken (and 2.8x runs at 0.5 FPS).

and don't even get me started on multiple linux installations on the same HDD ;)
`[[swap][boot][[OS1][OS2][OS3][...][data]][recovery]]`
where my programs are stored in the `data` partition (if not another HDD)
at least XFCE makes panel-launchers easy

**Comment by rubenwardy** [comment]
Closing in favour of #9166
